File: automator/src/main/java/com/stardust/automator/filter/TextFilter.java
package com.stardust.automator.filter;

import com.stardust.automator.UiObject;

import java.util.List;

/**
 * Created by Stardust on 2017/3/9.
 */

public class TextFilter extends ListFilter.Default {

    private static final KeyGetter TEXT_GETTER = new KeyGetter() {
        @Override
        public String getKey(UiObject nodeInfo) {
            CharSequence charSequence = nodeInfo.getText();
            return charSequence == null ? null : charSequence.toString();
        }

        @Override
        public String toString() {
            return "text";
        }
    };

    public static ListFilter equals(String text) {
        return new StringEqualsFilter(text, TEXT_GETTER);
    }

    public static ListFilter contains(String str) {
        return new StringContainsFilter(str, TEXT_GETTER);
    }

    public static ListFilter startsWith(String prefix) {
        return new StringStartsWithFilter(prefix, TEXT_GETTER);
    }

    public static ListFilter endsWith(String suffix) {
        return new StringEndsWithFilter(suffix, TEXT_GETTER);
    }

    public static ListFilter matches(String regex) {
        return new StringMatchesFilter(regex, TEXT_GETTER);
    }

    private String mText;

    private TextFilter(String text) {
        mText = text;
    }


    @Override
    public List<UiObject> filter(UiObject node) {
        return node.findByText(mText);
    }
}

File: automator/src/main/java/com/stardust/automator/filter/TextFilter.kt
package com.stardust.automator.filter

import com.stardust.automator.UiObject

/**
 * Created by Stardust on 2017/3/9.
 */

class TextFilter private constructor(private val mText: String) : ListFilter.Default() {


    override fun filter(node: UiObject): List<UiObject> {
        return node.findByText(mText)
    }

    companion object {

        private val TEXT_GETTER = object : KeyGetter {
            override fun getKey(nodeInfo: UiObject): String? {
                val charSequence = nodeInfo.text
                return charSequence?.toString()
            }

            override fun toString(): String {
                return "text"
            }
        }

        fun equals(text: String): ListFilter {
            return StringEqualsFilter(text, TEXT_GETTER)
        }

        fun contains(str: String): ListFilter {
            return StringContainsFilter(str, TEXT_GETTER)
        }

        fun startsWith(prefix: String): ListFilter {
            return StringStartsWithFilter(prefix, TEXT_GETTER)
        }

        fun endsWith(suffix: String): ListFilter {
            return StringEndsWithFilter(suffix, TEXT_GETTER)
        }

        fun matches(regex: String): ListFilter {
            return StringMatchesFilter(regex, TEXT_GETTER)
        }
    }
}
